src/amo/withExperiment.js

```diff
@@ -175,15 +181,79 @@ export const withExperiment =
         WrappedComponent,
       )})`;
 
-      componentDidMount() {
-        const { _getVariant, _isExperimentEnabled, cookies, id, variants } =
-          this.props;
-
+      initialExperimentSetup(props) {
+        const {
+          _getVariant,
+          _isExperimentEnabled,
+          dispatch,
+          id,
+          storedVariant,
+          variants,
+        } = props;
+
+        let { variant } = this;
         const isEnabled = _isExperimentEnabled({ _config, id });
         const registeredExperiments = this.getExperiments();
         const experimentInCookie = this.cookieIncludesExperiment(
           registeredExperiments,
         );
+        const addExperimentToCookie = !experimentInCookie;
+        const variantFromStore = storedVariant;
+
+        if (isEnabled && !variant) {
+          // Use the variant in the cookie if one exists, otherwise use the
+          // variant from the Redux store.
+          if (experimentInCookie) {
+            variant = registeredExperiments[id];
+          } else if (variantFromStore) {
+            variant = variantFromStore.variant;
+          }
+
+          // Do we need to store the variant in the cookie?
+          if (addExperimentToCookie) {
+            // Determine the variant if we don't already have one.
+            variant = variant || _getVariant({ variants }).id;
```

Comment:
This is quite confusing because a stored variant is `(id, variant)` but here `variant` can be either `variant` or `id`. I don't follow.

Reply:
I cleaned up the whole `StoredVariant` thing, but that isn't actually what is causing the problem here. The issue here is that `_getVariant()` doesn't return just the variant name (which is called `id`), but rather returns a `ExperimentVariant` object, which includes both the name/id and also the `percentage`. Looking at the code, however, I don't think we ever use that `percentage` anywhere. It's only really needed inside `getVariant`, so I could change `getVariant` so that it just returns the string id. I'll give that a shot.